### PR TITLE
Enable DAO Proposal to execute Arbitray Tx

### PR DESCRIPTION
**Pull request by DerTiedemann**
# Goals
Allow more or less (anything the implements a specific tx type) to be executed as a ConcludeProposalTx that will only be executed when the vote has passed. Also allow the Implementing Tx to verify the ProposalConfiguration to make sure it can reject parameters that are invalid.  E.g. not high enough threshold or invalid durration of Proposal.
# Non-Goals
Allow for multiple outcomes in the same proposal
# Implementation
Newly/Adjusted Tx:
- SubmitDaoProposalTx -> Creates and typechecks a proposal for a wrapped proposedTX (Currently only AddValidatorTx)
- AddDaoVoteTx -> Any currently active validator that has been a validator before the vote comenced can issue approval with this tx
- ConcludeDaoProposalTx -> is Issued automatically by the node and executes the proposed tx if the proposal was accepted

# Caveats
The proposedTx is checked at submit time, but the state can change while the proposal is active. If the proposalTx becomes invalid while the Proposal is invalid the proposal concludes as normal and ignores the failure. That might not be desireable but the easiest way of handeling it atm.

The AddValidator proposal will not be able to submited outside of a ConcludeProposalTx

### CODE REVIEW

File: vms/platformvm/block_builder.go [hunk 1]
```diff
@@ -306,6 +324,30 @@ func (m *blockBuilder) Shutdown() {
 	m.vm.ctx.Lock.Lock()
 }
 
+// getStakerToReward return the staker txID to remove from the primary network
```
Comment (knikos) [inline vms/platformvm/block_builder.go:327]: this documentation seems to have been copied from another method

File: vms/platformvm/client.go [hunk 2]
```diff
@@ -215,6 +215,34 @@ type Client interface {
 	GetBlock(ctx context.Context, blockID ids.ID, options ...rpc.Option) ([]byte, error)
 	// GetConfiguration returns genesis information of the primary network
 	GetConfiguration(ctx context.Context) (*GetConfigurationReply, error)
+	// // AddDaoProposal adds a new Proposal created by this nodeId.
```
Comment (knikos) [inline vms/platformvm/client.go:218]: hey @DerTiedemann , these were the commented out methods I was referring to..

File: vms/platformvm/dao/dao.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	return currentTime.After(d.EndTime())
+}
+
+// Verify validates the start / end of the proposal
+func (d *ProposalConfiguration) Verify() error {
+	if d.Start >= d.End {
+		return errStartTimeBeforeEndTime
+	}
+
+	if d.Thresh > 0 {
```
Comment (knikos) [inline vms/platformvm/dao/dao.go:78]: that should be rather d.Thresh <= 0

File: vms/platformvm/dao_add_vote_tx.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		case err != nil:
+			return nil, nil, err
+		case !currentTimestamp.Before(daoProposal.DaoProposalTx().EndTime()):
+			return nil, nil, fmt.Errorf("proposal: %s is not active anymore", tx.ProposalID.String())
+		case len(daoProposal.Votes()) >= int(daoProposal.DaoProposalTx().ProposalConfiguration.Thresh):
+			return nil, nil, fmt.Errorf("proposal: %s already accepted", tx.ProposalID.String())
+		case daoProposal.Voted(tx.NodeID):
+			return nil, nil, fmt.Errorf("node %s has already voted on proposal: %s", tx.NodeID.String(), tx.ProposalID.String())
+		case daoProposal.DaoProposalTx().StartTime().After(validator.AddValidatorTx().StartTime()):
+			return nil, nil, fmt.Errorf("node %s only became a validator after the proposal started and is thus not eligible to participate", tx.NodeID.String())
```
Comment (knikos) [inline vms/platformvm/dao_add_vote_tx.go:129]: That should rather be:
`daoProposal.DaoProposalTx().StartTime().**Before**(validator.AddValidatorTx().StartTime()):`

File: vms/platformvm/dao/dao.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+	// Unix time this Dao finishes
+	End uint64 `serialize:"true" json:"endTime"`
+}
+
+func HasConcluded(state ProposalState) bool {
+	return state == ProposalStateConcluded
+}
+
+// StartTime is the time that this validator will enter the validator set
```
Comment (knikos) [inline vms/platformvm/dao/dao.go:58]: this documentation is obviously copied from validator.go and does not describe its real purpose

**Comment by DerTiedemann** [comment]
Damm, i just noticed, that there is am implicit requirement for the Ins of both the Submit and WrappedTx to be disjunct. Which the current stake function does not care about... *sadge*

**Comment by DerTiedemann** [comment]
I converted i back to a draft as i found some bugs that are game breaking, ill need to discuss with @evlekht how to proceed.
Main Issue:
- At tx creation time you will need to reference two set of ins that are disjunct for the Submit & Wrapped Tx, the current implementaiton of spend does not do this and the service does also not take this into account which will lead to the Tx no being accepted. 

Ill need to see if there is a way to sidestep this issue.

**Comment by DerTiedemann** [comment]
superceded by https://github.com/chain4travel/caminogo/pull/269

**Pull request closed without merging**